utilities/titandb/titan_db_test.cc

```diff
@@ -361,6 +363,167 @@ TEST_F(TitanDBTest, DISABLED_ReadAfterDropCF) {
+  options_.env = mock_env.get();
+  Open();
+
+  std::map<std::string, std::string> data;
+  const int kNumEntries = 100;
+  for (uint64_t i = 1; i <= kNumEntries; i++) {
+    Put(i, &data);
+  }
+  ASSERT_EQ(kNumEntries, data.size());
+  CompactRangeOptions copts;
+  // no compaction to enable Flush
+  VerifyDB(data);
+
+  // test flush error handling
+  SyncPoint::GetInstance()->SetCallBack(
+    "FlushJob::Start", [&](void *) {
+      mock_env->SetFilesystemActive(false, Status::IOError("Injected error"));
+    });
+  SyncPoint::GetInstance()->EnableProcessing();
+  FlushOptions fopts;
+  ASSERT_TRUE(db_->Flush(fopts).IsIOError());
+  mock_env->SetFilesystemActive(true);
+  // subsequent writes return error too
+  WriteOptions wopts;
+  std::string key = "key_after_flush";
+  std::string value = "value_after_flush";
+  ASSERT_TRUE(db_->Put(wopts, key, value).IsIOError());
+  SyncPoint::GetInstance()->DisableProcessing();
+  SyncPoint::GetInstance()->ClearAllCallBacks();
+
+  // test compaction error handling
+  SyncPoint::GetInstance()->SetCallBack(
+    "BackgroundCallCompaction:0", [&](void *) {
+      mock_env->SetFilesystemActive(false, Status::IOError("Injected error"));
+    });
+  SyncPoint::GetInstance()->EnableProcessing();
+  ASSERT_TRUE(db_->CompactRange(copts, nullptr, nullptr).IsIOError());
+  mock_env->SetFilesystemActive(true);
+  // subsequent writes return error too
+  key = "key_after_compaction";
+  value = "value_after_compaction";
+  ASSERT_TRUE(db_->Put(wopts, key, value).IsIOError());
+  SyncPoint::GetInstance()->DisableProcessing();
+  SyncPoint::GetInstance()->ClearAllCallBacks();
+
+  options_.env = env_;
+  SyncPoint::GetInstance()->ClearAllCallBacks();
+  // env must be destructed AFTER db is closed to avoid
+  // `pure abstract method called` complaint.
+  Close();
+  db_ = nullptr;
+}
+
+TEST_F(TitanDBTest, BlobFileCorruptionErrorHandling) {
+  Open();
+  std::map<std::string, std::string> data;
+  const int kNumEntries = 100;
+  for (uint64_t i = 1; i <= kNumEntries; i++) {
+    Put(i, &data);
+  }
+  ASSERT_EQ(kNumEntries, data.size());
+  CompactRangeOptions copt;
+  ASSERT_OK(db_->CompactRange(copt, nullptr, nullptr));
+  VerifyDB(data);
+
+  // Modify the checksum data to reproduce a mismatch
+  SyncPoint::GetInstance()->SetCallBack(
+    "BlobDecoder::DecodeRecord", [&](void* arg) {
+      auto* crc = reinterpret_cast<uint32_t*>(arg);
+      *crc = *crc + 1;
+    });
+
+  SyncPoint::GetInstance()->EnableProcessing();
+  for (auto& it : data) {
+    std::string value;
+    if(it.second.size() < options_.min_blob_size) {
+        continue;
+    }
+    ASSERT_TRUE(db_->Get(ReadOptions(), it.first, &value).IsCorruption());
+  }
+
+  SyncPoint::GetInstance()->DisableProcessing();
+  std::unique_ptr<Iterator> iter(db_->NewIterator(ReadOptions()));
+  iter->SeekToFirst();
+  auto it = data.begin();
+  for (int i = 0; i < kNumEntries / 2 - 1; i++) {
+    ASSERT_TRUE(iter->Valid());
```

Comment:
The iterator should be invalid at i=1 (the first value with size > min_blob_size). Can you check?

Reply:
I was trying to simulate error `in the middle of iterating`, but at start is also fine.